## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [38]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [39]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [40]:
# Use the describe function to gather some basic statistics
print(df.describe())

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [41]:
# What three item categories had the most entries?
top_three = df['category'].value_counts().head(3)
print(top_three)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [42]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax()
print(f"Category with the most entries :{top_category}")
filtered_df = df[df['category'] == top_category]
subcategory_counts = filtered_df['subcategory'].value_counts()
top_subcategory = subcategory_counts.idxmax()
print(f"Subcategory with the most entries in the {top_category} category: {top_subcategory, subcategory_counts}")



Category with the most entries :consumables
Subcategory with the most entries in the consumables category: ('bathroom supplies', subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64)


In [43]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [44]:
# Store the client ids of those top 5 clients in a list.
top_clients_list = df['client_id'].value_counts().head(5).index.tolist()
print(top_clients_list)

[33615, 66037, 46820, 38378, 24741]


In [45]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = df['client_id'].value_counts().idxmax()  
print(f"Client with the most entries: {top_client}")
top_client_data = df[df['client_id'] == top_client]
total_units = top_client_data['qty'].sum()
print(f"Total units ordered by client {top_client}: {total_units}")

Client with the most entries: 33615
Total units ordered by client 33615: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [46]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
print(df[['unit_price', 'qty', 'line_subtotal']].head())

   unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00


In [53]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df['unit_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
print(df[['unit_weight', 'shipping_price']].head())

   unit_weight  shipping_price
0         7.50            75.0
1         1.49            14.9
2         1.68            16.8
3         1.23            12.3
4        46.43           464.3


In [48]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 0.0925
df['total_price'] = (df['line_subtotal'] + df['shipping_price']) * (1 + sales_tax_rate)
print(df[['line_subtotal', 'shipping_price','total_price']].head())

   line_subtotal  shipping_price    total_price
0      115164.00            75.0  125898.607500
1         523.95            14.9     588.693625
2         527.28            16.8     594.407400
3        1056.18            12.3    1167.314400
4        3902.00           464.3    4770.182750


In [49]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
print(df.head())




       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.

In [50]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_subtotal'] - df['line_cost']
print(df[['line_cost', 'line_subtotal', 'line_profit']])


       line_cost  line_subtotal  line_profit
0       80159.55      115164.00     35004.45
1         331.79         523.95       192.16
2         323.34         527.28       203.94
3         732.95        1056.18       323.23
4        2627.70        3902.00      1274.30
...          ...            ...          ...
54634    1725.30        2743.29      1017.99
54635    8363.62        9709.73      1346.11
54636   27263.85       31188.50      3924.65
54637     317.04         165.76      -151.28
54638    1680.61        3103.31      1422.70

[54639 rows x 3 columns]


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [23]:
# Check your work using the totals above
specific_order_id = 6128929
specific_order_total = df[df['order_id'] == specific_order_id]['total_price'].sum()
print(f"Total price for order {specific_order_id}: ${specific_order_total:.2f}")

Total price for order 6128929: $752748.75


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [51]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# Step 1: Group by client_id and calculate total quantity
client_quantity = (
    df.groupby('client_id')['qty']
    .sum()
    .reset_index()
    .sort_values('qty', ascending=False)
    .head(5)
)

# Step 2: Filter the DataFrame for the top 5 clients and calculate their total spending
top_client_spending = (
    df[df['client_id'].isin(client_quantity['client_id'])]
    .groupby('client_id')['total_price']
    .sum()
    .reset_index()
    .sort_values('total_price', ascending=False)
)

# Step 3: Round the total_price column for better presentation
top_client_spending['total_price'] = top_client_spending['total_price'].round(2)

# Step 4: Rename columns for clarity
top_clients_summary = top_client_spending.rename(columns={'total_price': 'total_spending'})

# Display the results
print(top_clients_summary)


   client_id  total_spending
2      78965    6.500970e+08
3      85931    2.263611e+08
4      86721    1.109862e+08
1      77354    4.592791e+07
0      30143    2.351848e+07


In [52]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
client_summary = df.groupby('client_id').agg(
    qty=('qty', 'sum'),                      
    shipping_price=('shipping_price', 'sum'), 
    line_price=('line_subtotal', 'sum'),        
    line_cost=('line_cost', 'sum'),           
    line_profit=('line_profit', 'sum')       
).reset_index()

# Step 2: Sort by qty and select the top 5 clients
top_5_clients = client_summary.sort_values(by='qty', ascending=False).head(5)

# Step 3: Round values for better presentation
top_5_clients = top_5_clients.round(2)

# Display the result
print(top_5_clients)

     client_id      qty  shipping_price    line_price     line_cost  \
758      78965  4033302          7383.3  5.950470e+08  4.032375e+08   
850      86721  1010426          6154.6  1.015831e+08  7.396203e+07   
204      30143   846069          2157.3  2.152505e+07  1.309872e+07   
840      85931   644423          2763.0  2.071928e+08  1.180031e+08   
738      77354   506382          2564.3  4.203671e+07  3.569827e+07   

      line_profit  
758  1.918096e+08  
850  2.762105e+07  
204  8.426335e+06  
840  8.918970e+07  
738  6.338442e+06  


In [59]:
# Format the data and rename the columns to names suitable for presentation.

# Step 1: Verify and update the money columns
money_columns = ['shipping_price', 'line_subtotal', 'line_cost', 'line_profit']

# Define a function that converts a dollar amount to millions
def currency_format_millions(amount):
    return amount / 1_000_000  # Convert to millions

# Step 2: Apply the currency_format_millions function to existing money columns
for col in money_columns:
    if col in df.columns:  # Check if column exists before applying
        df[col] = df[col].apply(currency_format_millions)
    else:
        print(f"Column '{col}' does not exist in the DataFrame.")

# Step 3: Rename the columns to reflect the change in the money format
df.rename(columns={
    'client_id': 'Client ID',
    'qty': 'Units',
    'shipping_price': 'Shipping (millions)',
    'line_subtotal': 'Total Revenue (millions)',
    'line_cost': 'Total Cost (millions)',
    'line_profit': 'Total Profit (millions)'
}, inplace=True)

# Step 4: Group by 'Client ID' and calculate totals for the top 5 clients
top_5_clients = (
    df.groupby('Client ID')
    .agg(
        Units=('Units', 'sum'),
        Shipping=('Shipping (millions)', 'sum'),
        Revenue=('Total Revenue (millions)', 'sum'),
        Cost=('Total Cost (millions)', 'sum'),
        Profit=('Total Profit (millions)', 'sum')
    )
    .reset_index()
    .sort_values(by='Profit', ascending=False)  # Sort by profit
    .head(5)
)

# Step 5: Round values for presentation
top_5_clients = top_5_clients.round(6)

# Display the results
print(top_5_clients)

Column 'shipping_price' does not exist in the DataFrame.
Column 'line_subtotal' does not exist in the DataFrame.
Column 'line_cost' does not exist in the DataFrame.
Column 'line_profit' does not exist in the DataFrame.
     Client ID    Units  Shipping     Revenue        Cost      Profit
758      78965  4033302  0.007383  595.047042  403.237454  191.809588
840      85931   644423  0.002763  207.192784  118.003086   89.189698
144      24741   239862  0.009366   70.176885   40.571817   29.605068
320      38473   402770  0.007832  163.620504  135.817674   27.802830
850      86721  1010426  0.006155  101.583080   73.962028   27.621051


In [60]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_data = top_5_clients.sort_values(by='Total Profit (millions)', ascending=False)
sorted_by_profit = sorted_data
print(sorted_by_profit)
                                        

KeyError: 'Total Profit (millions)'